### ISSUE TITLE
How to just download binary dependencies

### ISSUE THREAD

**Issue by KoCMoHaBTa** [Carthage/Carthage#2229]
* `carthage version`: 0.26.2
* `xcodebuild -version`: Xcode 9.0.1
* Are you using `--no-build`? YES
* Are you using `--no-use-binaries`? NO
* Are you using `--use-submodules`? NO
* Are you using `--cache-builds`? NO

Hi, 

For git dependencies, I use `carthage update --no-build` and `carthage checkout` commands in order to respectively update and resolve dependencies.
Then i manually integrate the downloaded projects as references from the `Carthage/Checkouts` directory. For more information, you can check my [simple project](https://github.com/KoCMoHaBTa/CarthageIntegrationSample) 

This works fine, however i need to just download some 3rd party prebuilt frameworks as they are and integrate them manually.
More specifically, there is a zip file that contains multiple frameworks and resources that i just need to download, unzip and manually integrate as usually, giving me the benefit of avoiding copying the frameworks and resources into my repository.
Also i'm looking for a way to manage just resource dependencies in the same way.

I've tried to define the dependency as `binary`, however Carthage appears to do some post processing that fails and never presents the downloaded files. The error is `the DTSDKName key in its plist file is missing`, however i just want Carthage to download the zip file and extract it - then i'll do the integration with its contents myself, no matter what they are.

My question is - is there an existing command that can fulfil my needs, or if not would it be possible to introduce such?

**Comment by blender** [comment]
> there is a zip file that contains multiple frameworks and resources that i just need to download

Why would carthage download a zip file it knows nothing about 

> The error is the DTSDKName key in its plist file is missing

and that is not a framework archive ?

**Comment by KoCMoHaBTa** [comment]
> Why would carthage download a zip file it knows nothing about
For the same reason it can just download any git repository, no matter what kind of code is inside.
What i'm suggesting is to have an additional option that would allow downloading binary without inspecting and copying them in some specific way, as it is now per platform. Similar to `--no-build` - you get the source code and thats it - we miss the behaviour to just download the frameworks.

For more information - the frameworks are static and probably that is the reason for the error.
But Carthage is totally capable of delivering the contents of the zip file, but it does not do it.

The archive is not owned by me, its 3rd party closed source and i have no control over how it is structured.

**Comment by blender** [comment]
I understand. I suggest you use nix for this instead.

**Comment by KoCMoHaBTa** [comment]
Ok, let me put it this way - if I contribute to develop this behaviour as an additional option to existing commands, would it be accepted?

**Comment by blender** [comment]
Not for me to say. I think direction here is needed. @ikesyo @mdiep .

Change is required somewhere around here https://github.com/Carthage/Carthage/blob/master/Source/CarthageKit/Project.swift#L824 along with a new CLI switch

**Comment by ikesyo** [comment]
I think arbitrary binary downloading feature is out of scope.

**Comment by mdiep** [comment]
> I've tried to define the dependency as binary, however Carthage appears to do some post processing that fails and never presents the downloaded files. The error is the DTSDKName key in its plist file is missing, however i just want Carthage to download the zip file and extract it - then i'll do the integration with its contents myself, no matter what they are

I think Carthage is trying to do what you want. But when we see a `.framework`, we need to know its platform (so we can know whether to put it in `Carthage/Build/iOS/` or `Carthage/Build/Mac`, e.g.). It sounds like the framework you're trying to use lacks the information to make this decision.

I'd suggest digging into why Carthage wants the `DTSDKName`, how it uses it, why your framework doesn't have it, and whether we can use other information (that your framework _does_ have) to determine the same thing.

**Comment by KoCMoHaBTa** [comment]
What does out of scope means?
Let say that i build this feature as an additional command option, eg `-copy-binaries` (or suggest another name) to the `checkout` command - will it be accepted as an optional behaviour?

Carthage is already doing this - it is about providing the downloaded contents and not performing the platform validation.

**Comment by mdiep** [comment]
I think it'd be reasonable to make `binary` support static frameworks—since we've been expanding support for them generally in light of Xcode's increased support. That's how I'd suggest that you proceed: use the existing functionality, but make it work for static frameworks as well.

**Comment by blender** [comment]
I understand @KoCMoHaBTa  has 

> More specifically, there is a zip file that contains multiple frameworks and resources

Not sure how that would work

**Comment by KoCMoHaBTa** [comment]
Image it in the following way:
- no default behaviour will be changed
- if you pass `--copy-binaries` argument to the `checkout` command - this will produce the following folder `<project_dir>/Carthage/Downloads` that will contain the extracted zip contents of each downloaded binary dependency. 
- if you pass `--no-copy-frameworks` argument to the `checkout` command - this will skip the framework validation and will not copy them to the `<project_dir>/Carthage/Build` folder.

So basically to just download a zip file as a binary dependency - you have to pass `--copy-binaries` and `--no-copy-frameworks` arguments.

This way - we can manually integrate with static frameworks and their resources.

**Comment by KoCMoHaBTa** [comment]
Example of such frameworks are Fabric / Crashlytics and Firebase.
Having this ability and defining the appropriate JSON configuration for the binary dependency - we can use Carthage to resolve them.

**Comment by KoCMoHaBTa** [comment]
Here is a PR that address the issue i'm talking about.

**Comment by KoCMoHaBTa** [comment]
Closing due to inability and unwillingness of the owner to understand the proposed feature and benefits of supporting any kind of binary dependency.

**Comment by mdiep** [comment]
Don't try to put that on me.

I've demonstrated a willingness to try to understand the problem your facing and to help you come up with an acceptable solution.

Instead of saying "here are the specifics of the situation I'm in, here's how I think Carthage could meet it, let's work together to find a good solution", you've repeatedly presented a specific vision without giving the proper context.

I suggest you take a look at your own communication and sense of entitlement. If you'd like to start from a problem statement and work to find a solution, then please open a new issue.
